### ISSUE TITLE
Random GetBucketLocation Failures Since Upgrading to v0.9.19

### ISSUE THREAD

**Issue by KISStian** [cloud-custodian/cloud-custodian#7870]
### Describe the bug

While testing v0.9.18 we encountered the AccessDenied error for the GetBucketLocation api while policies were trying to write to a central S3 logging bucket in one region.  After realizing this was a known issue for which a fix was being delivered in v0.9.19, we held off on our upgrade from v0.9.14.  We did testing in our PreProd environment when v0.9.19 was released and the issue seemed to have been resolved.

Two nights ago we moved forward with the upgrade and since then we have been receiving random failures.  We have several dozen policies running against an Organization of 500+ accounts across 16 regions.  Pull-type policies run hourly and daily, and this is where we are seeing the random errors.  Event-based policies that run in lambda do not appear to be experiencing this, st least we haven't identified any occurrences of it yet.

### What did you expect to happen?

The policies would run successfully and not receive AccessDenied failures for the GetBucketLocation api.

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

```shell
Custodian:   0.9.19
Python:      3.9.10 (main, Jan 15 2022, 11:48:04) 
             [Clang 13.0.0 (clang-1300.0.29.3)]
Platform:    posix.uname_result(sysname='Darwin', nodename='PL1USCLT001MAC.local', release='21.6.0', version='Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64', machine='x86_64')
Using venv:  True
Docker: False
Installed: 

PyYAML==6.0
Pygments==2.13.0
argcomplete==2.0.0
attrs==22.1.0
aws-xray-sdk==2.10.0
bleach==5.0.1
boto3==1.24.87
botocore==1.27.87
c7n==0.9.19
cachetools==5.2.0
certifi==2022.9.24
charset-normalizer==2.1.1
click==8.1.3
colorama==0.4.5
coverage==6.5.0
docutils==0.17.1
execnet==1.9.0
flake8==3.9.2
freezegun==1.2.2
google-api-core==2.10.1
google-api-python-client==2.64.0
google-auth==2.12.0
google-auth-httplib2==0.1.0
google-cloud-appengine-logging==1.1.5
google-cloud-audit-log==0.2.4
google-cloud-core==2.3.2
google-cloud-logging==3.2.4
google-cloud-monitoring==2.11.2
google-cloud-storage==1.44.0
google-crc32c==1.5.0
google-resumable-media==2.4.0
googleapis-common-protos==1.56.4
grpc-google-iam-v1==0.12.4
grpcio==1.49.1
grpcio-status==1.49.1
httplib2==0.20.4
idna==3.4
importlib-metadata==4.13.0
importlib-resources==5.9.0
iniconfig==1.1.1
jaraco.classes==3.2.3
jmespath==1.0.1
jsonpatch==1.32
jsonpointer==2.3
jsonschema==4.16.0
keyring==23.9.3
mccabe==0.6.1
mock==4.0.3
more-itertools==8.14.0
multidict==6.0.2
packaging==21.3
pkginfo==1.8.3
pkgutil-resolve-name==1.3.10
placebo==0.9.0
pluggy==1.0.0
portalocker==2.5.1
proto-plus==1.22.1
protobuf==4.21.7
psutil==5.9.2
py==1.11.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.7.0
pyflakes==2.3.1
pygments==2.13.0
pyparsing==3.0.9
pyrsistent==0.18.1
pytest==7.1.3
pytest-cov==3.0.0
pytest-forked==1.4.0
pytest-recording==0.12.1
pytest-sugar==0.9.5
pytest-terraform==0.6.4
pytest-xdist==2.5.0
python-dateutil==2.8.2
pyyaml==6.0
ratelimiter==1.2.0.post0
readme-renderer==37.2
requests==2.28.1
requests-toolbelt==0.9.1
retrying==1.3.3
rfc3986==2.0.0
rsa==4.9
s3transfer==0.6.0
six==1.16.0
tabulate==0.8.10
termcolor==2.0.1
tomli==2.0.1
tqdm==4.64.1
twine==3.8.0
typing-extensions==4.3.0
uritemplate==4.1.1
urllib3==1.26.12
vcrpy==4.2.1
webencodings==0.5.1
wrapt==1.14.1
yarl==1.8.1
zipp==3.8.1
```

### Policy

```shell
The failure is not tied to a specific policy.
```

### Relevant log/traceback output

```shell
2022-10-12 13:20:03,574: c7n_org:ERROR Exception running policy:ec2-optin-start-instances-off-hours-periodic account:XXXXXX region:us-east-2 error:unable to determine a region for output bucket XXX-XXX-XXX: An error occurred (AccessDenied) when calling the GetBucketLocation operation: Access Denied
```

**Comment by KISStian** [comment]
I just wanted to add that the policies are running in CodeBuild.

**Comment by ajkerrigan** [comment]
Thanks for reporting this. The intermittent nature makes it trickier to nail down, but if you're getting access denied on `GetBucketLocation` it means that finding the region via unauthenticated HTTP requests is _sometimes_ not working (and I suspect that leads to `GetBucketLocation` _always_ not working when the code makes it that far).

A couple things I can think to do here:
* **Get more info:** If we hit an HTTP error and can't determine a region, at least emit a warning to make it clearer in the logs when we'll need to fall back to `GetBucketLocation`
* **Cache bucket regions:** Cache the results of `get_bucket_region_clientless`, because:
  * Once we find a bucket's region, it's not going to change
  * One possible cause of this intermittent issue is that repeat HTTP requests eventually start getting throttled (this will be clearer if we **get more info**) - caching would avoid that, and seems like a reasonable thing to do in any case

Heads up to @PratMis - since you helped test #7682 it's worth having this on your radar in case you see similar intermittent issues.

**Comment by KISStian** [comment]
I am not sure if this is at all related (and if not I can open up a separate issue) but since upgrading we have experienced a few random TCP errors like the following:

`2022-10-12 21:24:10,044: c7n_org:ERROR Exception running policy:ec2-remediate-stop-ssm-agent-offline-periodic account:XXXXXXXXX region:us-west-2 error:<urlopen error [Errno 104] Connection reset by peer>`

These occur far less than the AccessDenied errors, but we never experienced them prior to upgrading to v0.9.19

<img width="965" alt="image" src="https://user-images.githubusercontent.com/35926785/195452118-a1eee8d0-7052-4810-ac10-3d4eb7024c7b.png">

**Comment by PratMis** [comment]
Hi @ajkerrigan , looks like in some of our account we're also seeing some new tracebacks. They are intermittent 
```
Traceback (most recent call last):
  File "/usr/lib/python3.7/urllib/request.py", line 1324, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/usr/lib/python3.7/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.7/http/client.py", line 1030, in _send_output
    self.send(msg)
  File "/usr/lib/python3.7/http/client.py", line 970, in send
    self.connect()
  File "/usr/lib/python3.7/http/client.py", line 1426, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/commands.py", line 302, in run
    policy()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 1297, in __call__
    resources = mode.run()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 291, in run
    with self.policy.ctx as ctx:
  File "/usr/local/lib/python3.7/dist-packages/c7n/ctx.py", line 74, in __enter__
    self.initialize()
  File "/usr/local/lib/python3.7/dist-packages/c7n/ctx.py", line 48, in initialize
    self.output = blob_outputs.select(self.options.output_dir, self)
  File "/usr/local/lib/python3.7/dist-packages/c7n/output.py", line 54, in select
    return self[k](ctx, parse_url_config(selector))
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/aws.py", line 607, in __init__
    get_bucket_region_clientless(self.bucket, s3_client.meta.endpoint_url) or
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/aws.py", line 165, in get_bucket_region_clientless
    response = urlopen(request)  # nosec B310
  File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 1367, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/usr/lib/python3.7/urllib/request.py", line 1326, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 0] Error>
```
This one started with the new release. Haven't seen this before
```
2022-10-13 05:23:48,872: custodian.output:ERROR Error while executing policy
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 302, in run
    resources = self.policy.resource_manager.resources()
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 534, in resources
    resources = self.augment(resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 590, in augment
    return self.source.augment(resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/appelb.py", line 44, in augment
    self.manager.retry)
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/appelb.py", line 111, in _describe_appelb_tags
    list(w.map(_process_tags, chunks(albs, 20)))
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 587, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 587, in <listcomp>
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 172, in submit
    self._adjust_thread_count()
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 193, in _adjust_thread_count
    t.start()
  File "/usr/lib/python3.7/threading.py", line 852, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
2022-10-13 05:23:48,874: custodian.output.blob:DEBUG s3: uploading policy logs
2022-10-13 05:23:48,876: custodian.commands:ERROR Error while executing policy appelb-XXX-tagging-XXXX, continuing
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 302, in run
    resources = self.policy.resource_manager.resources()
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 534, in resources
    resources = self.augment(resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 590, in augment
    return self.source.augment(resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/appelb.py", line 44, in augment
    self.manager.retry)
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/appelb.py", line 111, in _describe_appelb_tags
    list(w.map(_process_tags, chunks(albs, 20)))
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 587, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 587, in <listcomp>
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 172, in submit
    self._adjust_thread_count()
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 193, in _adjust_thread_count
    t.start()
  File "/usr/lib/python3.7/threading.py", line 852, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/commands.py", line 302, in run
    policy()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 1297, in __call__
    resources = mode.run()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 348, in run
    return resources
  File "/usr/local/lib/python3.7/dist-packages/c7n/ctx.py", line 103, in __exit__
    self.output.__exit__(exc_type, exc_value, exc_traceback)
  File "/usr/local/lib/python3.7/dist-packages/c7n/output.py", line 523, in __exit__
    self.upload()
  File "/usr/local/lib/python3.7/dist-packages/c7n/output.py", line 533, in upload
    self.upload_file(os.path.join(root, f), key)
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/aws.py", line 621, in upload_file
    'ServerSideEncryption': 'AES256'})
  File "/usr/local/lib/python3.7/dist-packages/boto3/s3/transfer.py", line 285, in upload_file
    filename, bucket, key, extra_args, subscribers
  File "/usr/local/lib/python3.7/dist-packages/s3transfer/manager.py", line 334, in upload
    call_args, UploadSubmissionTask, extra_main_kwargs
  File "/usr/local/lib/python3.7/dist-packages/s3transfer/manager.py", line 531, in _submit_transfer
    main_kwargs=main_kwargs,
  File "/usr/local/lib/python3.7/dist-packages/s3transfer/futures.py", line 474, in submit
    future = ExecutorFuture(self._executor.submit(task))
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 172, in submit
    self._adjust_thread_count()
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 193, in _adjust_thread_count
    t.start()
  File "/usr/lib/python3.7/threading.py", line 852, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
```
Could be related to exhausting the number of threads re: HTTP request to get s3 bucket region

**Comment by KISStian** [comment]
We are still reviewing the errors and trying to see if there is some pattern to what is occurring, especially considering the majority of the errors are in two regions.  While looking at the data, the high volume errors occur at the same time each day during our hourly c7n processing:
<img width="1672" alt="image" src="https://user-images.githubusercontent.com/35926785/196774914-5ef58e59-98a6-42b0-8519-78c97cb38ad2.png">
I found this to be quite odd.  Anyhow, I think that the idea of caching the bucket location is a great idea as that doesn't change....or just reverting the logic so that it doesn't even try to determine the region :)

**Comment by PratMis** [comment]
Hi @KISStian, we were also running into a bunch of errors, although different from what you have mentioned. `botocore.exceptions.CredentialRetrievalError` and `RuntimeError: can't start new thread` were the most common ones for us which could fix via https://github.com/uswitch/kiam/issues/96#issuecomment-417315210

**Comment by ajkerrigan** [comment]
Thanks @KISStian and @PratMis, your detailed info is helpful in sorting out what's going on and how we can handle it. In an earlier comment I mentioned that better logging (to help highlight these cases) and caching (to reduce the call volume and eliminate some classes of problems) seemed like useful ways forward.

Your logs and screenshots seem to support those changes, since there's some variety in the failures but call volume seems to be a factor across the board. Let me get a change together for this.

**Comment by ajkerrigan** [comment]
I haven't been able to reproduce these issues, probably because it depends on a specific set of environmental factors combined with resource/policy scale that I don't have handy.

I _have_ been able to confirm that when running a set of 65 policies using an output bucket in another region, #7923 reduces the number of HTTP HEAD requests from 65 to 1. That causes the policies to finish more quickly and _should_ avoid or at least indicate more clearly in logs when this issue pops up.

**Comment by KISStian** [comment]
Sounds great.  Unless you see any issues with this, I would like to take the updated aws.py module with your changes and place it into our runtime environment.

**Comment by KISStian** [comment]
Unfortunately, I hit the following error while running with the updated module:

`022-10-25 15:56:31,337: custodian.aws:WARNING unable to determine output bucket region with HTTP HEAD request: HTTP Error 503: Slow Down`

I received a few of those error in the logs.

**Comment by kapilt** [comment]
@KISStian fwiw this should get addressed with #8289

**Comment by ajkerrigan** [comment]
Closing this since the issues related to `GetBucketLocation` failures and HTTP request spamming should be resolved at this point.
